README.md
# mailform
Send out a given mail via a web form.

Allows sending mails to a configured and given mail address.

## Badges - remote integrations

[![Build Status](https://travis-ci.org/ottlinger/mailform.svg?branch=master)](https://travis-ci.org/ottlinger/mailform)

[![codecov](https://codecov.io/gh/ottlinger/mailform/branch/master/graph/badge.svg)](https://codecov.io/gh/ottlinger/mailform)

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/07b670b13aa944a789f40bbdf297b337)](https://www.codacy.com/app/github_25/mailform?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ottlinger/mailform&amp;utm_campaign=Badge_Grade)

[![Waffle.io - Columns and their card count](https://badge.waffle.io/ottlinger/mailform.svg?columns=all)](https://waffle.io/ottlinger/mailform)
as Waffle is about to disappear in 2 months: [Github Project Mailform](https://github.com/users/ottlinger/projects/1?add_cards_query=is%3Aopen)

[![Apache v2.0](https://img.shields.io/github/license/ottlinger/mailform.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)

## PHP project init
### PHP

Project requires at least PHP 7.2.

#### Ubuntu
Depending on your installation you may need to install a more recent PHP version,
as I had to do on Ubuntu 18.0.4 LTS:
https://tecadmin.net/install-php-7-on-ubuntu/

* Upgrade to PHP 7.2
```
sudo apt-get install -y php7.2
```
* Additional modules:
```
sudo apt install php7.2-xml php7.2-mbstring php7.2-mysql php7.2-intl php7.2-sqlite3 php-xdebug sqlite3
```
#### Mac
```
$ brew install php@7.2
```
In case you have older versions of PHP installed follow this cleanup procedure:
https://medium.com/@romaninsh/install-php-7-2-xdebug-on-macos-high-sierra-with-homebrew-july-2018-d7968fe7e8b8

If you have multiple PHP versions and need to switch per project you may need this:
```
$ brew install brew-php-switcher
```

### Composer
```
php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
php -r "if (hash_file('sha384', 'composer-setup.php') === '48e3236262b34d30969dca3c37281b3b4bbe3221bda826ac6a9a62d6444cdb0dcd0615698a5cbe587c3f0fe57a54d8f5') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;"
php composer-setup.php
php -r "unlink('composer-setup.php');"
```

done with version 1.8.4.

### Project init

```
$ ./composer.phar init
```

#### Project layout

tbd: https://getcomposer.org/doc/01-basic-usage.md#introduction

##### Base structure

following https://blog.nikolaposa.in.rs/2017/01/16/on-structuring-php-projects/

```
bin/              # command-line executables
config/           # configuration files
public/           # web server files, assets
resources/        # other resource files
src/              # PHP source code
templates/        # view and layout files
tests/            # test code
```

#### PHPUnit setup

https://phpunit.de/getting-started/phpunit-8.html

```
$ composer require --dev phpunit/phpunit ^8 (needs a modern PHP)
```

#### PHPMailer integration

In order to integrate PHPMailer you may want to run:
```
$ composer require phpmailer/phpmailer
```

#### Templates

The mail form web template is based on [Html5UP ReadOnly](https://html5up.net/read-only/download)
which is available under [Creative Commons Attribution 3.0 Unported](./templates/LICENSE.txt)

#### Favicon

The used favicon [House](https://www.freefavicon.com/freefavicons/objects/iconinfo/house-152-237998.html) is taken from [FreeFavicon](http://www.freefavicon.com/blog/).

#### Application configuration

This application can be configured if used in you own context.
Please copy the given template [mailform-config.php.template](./config/mailform-config.php.template)
to [mailform-config.php](./config) in the config-directory.

#### Docker integration

With a more recent Docker setup you may simply run:
```
$ docker-compose up
```

If you see something like:
```
Recreating mailform_php-apache_1 ... done
Attaching to mailform_php-apache_1
....
```
you are ready to visit http://localhost:8888
to see the application up and running.

In case you need to tweak PHP settings:
http://localhost:8888/phpinfo.php
might be of use - please remove this file in production deployments!

* https://linuxconfig.org/how-to-create-a-docker-based-lamp-stack-using-docker-compose-on-ubuntu-18-04-bionic-beaver-linux
* https://hub.docker.com/_/php
* https://docs.docker.com/install/linux/docker-ce/ubuntu/

HINT! Even if you enable mail sending this will not work from within the container - you need to deploy the application first :-)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Docker support for local development
- PHPUnit integration
- Composer project setup
- Make mail body configurable
### Changed
- Enhanced configurability of application with application config template found under [config](./config)
### Removed

## [0.0.1] - 2019-xx-yy
tbd

[Unreleased]: https://github.com/ottlinger/mailform/compare/v1.0.0...HEAD
[0.0.2]: https://github.com/ottlinger/mailform/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/ottlinger/mailform/releases/tag/v0.0.1


templates/index.php
<!DOCTYPE HTML>
<!--
	Read Only by HTML5 UP
	html5up.net | @ajlkn
	Free for personal and commercial use under the CCA 3.0 license (html5up.net/license)
-->
<?php
require_once __DIR__ . '/../vendor/autoload.php';

use mailform\FormHelper;
use mailform\Mailer;
use mailform\Message;

$hasErrors = false;
$isRobot = !(boolval(FormHelper::isSetAndNotEmptyInArray($_POST, "mailform-priority")) && "middle" === FormHelper::filterUserInput($_POST['mailform-priority']));
$sendOut = false;

?>
<html lang="en">
<head>
    <title>Mailform - example application</title>
    <meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no"/>
    <link rel="stylesheet" href="assets/css/main.css"/>

    <link rel="icon" type="image/x-icon" href="assets/favicon/favicon.ico">
    <link rel="apple-touch-icon-precomposed" href="assets/favicon/house-152-237998.png">
    <meta name="msapplication-TileColor" content="#ffffff"/>
    <meta name="msapplication-TileImage" content="assets/favicon/house-144-237998.png">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="assets/favicon/house-152-237998.png">
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="assets/favicon/house-144-237998.png">
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="assets/favicon/house-120-237998.png">
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="assets/favicon/house-114-237998.png">
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="assets/favicon/house-72-237998.png">
    <link rel="apple-touch-icon-precomposed" href="assets/favicon/house-57-237998.png">
    <link rel="icon" href="assets/favicon/house-32-237998.png" sizes="32x32">
</head>
<body class="is-preload">

<!-- Header -->
<section id="header">
    <header>
        <span class="image avatar"><img src="images/avatar.jpg" alt=""/></span>
        <h1 id="logo"><a href="#">Mailform</a></h1>
        <p>Send out mails</p>
    </header>
    <nav id="nav">
        <ul>
            <li><a href="#contact" class="active">Contact form</a></li>
        </ul>
    </nav>
    <footer>
        <ul class="icons">
            <li>Feedback via <a href="https://www.github.com/ottlinger/mailform" target="_blank" class="icon fa-github"><span
                        class="label">Github</span></a></li>
            <li><a href="https://aiki-it.de" target="_blank">AIKI IT</a></li>
        </ul>
    </footer>
</section>

<!-- Wrapper -->
<div id="wrapper">

    <!-- Main -->
    <div id="main">

        <!-- Contact form -->
        <section id="contact">
            <div class="container">
                <h3>Mailform example application</h3>
                <p>This mail form allows to send a message to a configured recipient address and the submitted mail
                    address. Apart from that there is a little
                    spam protection available.</p>
                <?php
                if (boolval(Mailer::getFromConfiguration("sendmails"))) {
                    print "<h1 style=\"color:green;\">Application is configured to really send out mails!</h1>";
                } else {
                    print "<h1 style=\"color:lawngreen;\">Application is running in demo-mode and will not send out mails.</h1>";
                }

                if ($_SERVER['REQUEST_METHOD'] === 'POST') {
                    $message = new Message($_POST['mailform-name'], $_POST['mailform-message'], $_POST['mailform-email']);
                    $mailer = new Mailer($message, true);

                    if (!$message->isValid() || $isRobot) {
                        print "<h4 style=\"color:red;\">There were errors while submitting the form, please provide all mandatory fields and a valid email.</h4>";
                        $hasErrors = true;
                    } else {
                        $mailer->sendAllMails();
                        $sendOut = true;
                    }
                }

                if ($sendOut) {
                    print "<h4>Thanks for submitting your request at " . date('Y-m-d H:i:s') . "</h4>";
                    print "<div class=\"col-12\"><p>You may return to our >>> <a href=\"" . Mailer::getFromConfiguration('successlinktarget') . "\">main application page</a></p></div>";
                }
                ?>

                <?php
                if ($_SERVER['REQUEST_METHOD'] != 'POST' || $hasErrors) {
                ?>
                <form method="post" action="#">
                    <?php
                    } // end if not POST
                    ?>
                    <div class="row gtr-uniform">
                        <!-- TODO generify: all mailform-something ids are selected and put into the mail -->
                        <!--div class="col-12">
                            <label for="mailform-category">Please select the category of your request:</label>
                            <select name="mailform-category" id="mailform-category">
                                <option value="1">Contact request</option>
                                <option value="2">Book request</option>
                                <option value="3">Administration</option>
                            </select>
                        </div-->
                        <div class="col-6 col-12-xsmall"><input type="text" name="mailform-name" id="mailform-name"
                                <?php
                                if ($hasErrors || boolval(FormHelper::isSetAndNotEmptyInArray($_POST, "mailform-name"))) {
                                    print " value='" . FormHelper::filterUserInput($_POST['mailform-name']) . "' ";
                                    if ($message->hasNameErrors()) {
                                        print " style=\"border-color: red;\" ";
                                    }
                                }
                                ?>
                                                                placeholder="Your name"/>
                        </div>
                        <div class="col-6 col-12-xsmall"><input type="email" name="mailform-email" id="mailform-email"
                                <?php
                                if ($hasErrors || boolval(FormHelper::isSetAndNotEmptyInArray($_POST, "mailform-email"))) {
                                    print " value='" . FormHelper::filterUserInput($_POST['mailform-email']) . "' ";
                                    if ($message->hasMailErrors()) {
                                        print " style=\"border-color: red;\" ";
                                    }
                                }
                                ?>

                                                                placeholder="Your email"/></div>
                        <div class="col-12">
                            <label for="mailform-message">We are looking forward to your message:</label>
                            <textarea name="mailform-message" id="mailform-message" placeholder="Your message"
                                <?php
                                if (isset($message) && $message->hasContentsErrors()) {
                                    print " style=\"border-color: red;\" ";
                                }
                                ?>
                                      rows="6"><?php
                                if ($hasErrors || boolval(FormHelper::isSetAndNotEmptyInArray($_POST, "mailform-message"))) {
                                    print FormHelper::filterUserInput($_POST['mailform-message']);
                                }
                                ?></textarea></div>

                        <?php
                        if (!$sendOut) {
                            ?>

                            <p
                                <?php
                                if ($isRobot && $_SERVER['REQUEST_METHOD'] === 'POST') {
                                    print " style=\"color: red;\" ";
                                };
                                ?>
                            >Please click the <strong>middle</strong> button, in order to proof that you are not a
                                robot:
                            </p>

                            <div class="col-4 col-12-medium">
                                <input type="radio" id="mailform-priority-low" name="mailform-priority" value="low"
                                    <?php
                                    if (boolval(FormHelper::isSetAndNotEmptyInArray($_POST, "mailform-priority")) && "low" === FormHelper::filterUserInput($_POST['mailform-priority'])) {
                                        print ' checked';
                                    }
                                    ?>
                                >
                                <label for="mailform-priority-low">Left</label>
                            </div>
                            <div class="col-4 col-12-medium">
                                <input type="radio" id="mailform-priority-normal" name="mailform-priority"
                                       value="middle"
                                    <?php
                                    if (boolval(FormHelper::isSetAndNotEmptyInArray($_POST, "mailform-priority")) && "middle" === FormHelper::filterUserInput($_POST['mailform-priority'])) {
                                        print ' checked';
                                    }
                                    ?>
                                >
                                <label for="mailform-priority-normal">Middle</label>
                            </div>
                            <div class="col-4 col-12-medium">
                                <input type="radio" id="mailform-priority-high" name="mailform-priority" value="high"
                                    <?php
                                    if (boolval(FormHelper::isSetAndNotEmptyInArray($_POST, "mailform-priority")) && "high" === FormHelper::filterUserInput($_POST['mailform-priority'])) {
                                        print ' checked';
                                    }
                                    ?>
                                >
                                <label for="mailform-priority-high">Right</label>
                            </div>

                            <?php
                        } // end if DO NOT SHOW if sendOut

                        if ((!$hasErrors && $_SERVER['REQUEST_METHOD'] != 'POST') || $hasErrors) {
                            ?>
                            <div class="col-12">
                                <ul class="actions">
                                    <li><input type="submit" class="button primary" value="Send Message"/></li>
                                    <li><input type="reset" value="Reset Form"/></li>
                                </ul>
                            </div>
                            <?php
                        } // end if NOT-POST
                        if ($_SERVER['REQUEST_METHOD'] === 'POST' && boolval(Mailer::getFromConfiguration("debug"))) {
                            print "<code class='col-12'>Data that was submitted: ";
                            var_dump($_POST);
                            print "</code>";
                        }
                        ?>
                    </div>
                </form>
            </div>
        </section>

    </div>

    <!-- Footer -->
    <section id="footer">
        <div class="container">
            <ul class="copyright">
                <li>&copy; Mailform, <?php print date("Y-m-d"); ?> All rights reserved.</li>
                <li>Design: <a href="https://html5up.net">HTML5 UP</a></li>
                <li>Served by <a href="https://www.github.com/ottlinger/mailform">Mailform</a> from <a
                        href="https://aiki-it.de">AIKI IT</a></li>
            </ul>
        </div>
    </section>
</div>

<!-- Scripts -->
<script src="assets/js/jquery.min.js"></script>
<script src="assets/js/jquery.scrollex.min.js"></script>
<script src="assets/js/jquery.scrolly.min.js"></script>
<script src="assets/js/browser.min.js"></script>
<script src="assets/js/breakpoints.min.js"></script>
<script src="assets/js/util.js"></script>
<script src="assets/js/main.js"></script>

</body>
</html>
